GoodThingsComein3s/API/APIManager.m

```diff
@@ -44,36 +48,104 @@ - (NSMutableURLRequest*)_getURLRequestForURLString: (NSString *)urlString {
-               completion(restaurants,nil);
+               NSArray *restaurants = [Restaurant restaurantsWithArray:restaurantDictionaries];
+               
+               completion([self filterRestaurants:restaurants],nil);
            }
     }];
     [task resume];
+    
+}
+
+- (NSArray*)filterRestaurants: (NSArray*)restaurantsToFilter {
+    for(Restaurant *restaurant in restaurantsToFilter){
+        restaurant.score = [self calculateRestaurantScore:restaurant];
+    }
+    restaurantsToFilter = [restaurantsToFilter sortedArrayUsingComparator:^NSComparisonResult(Restaurant *restaurant1, Restaurant *restaurant2) {
+        if (restaurant1.score.intValue < restaurant2.score.intValue) {
+            return NSOrderedDescending;
+        } else if (restaurant1.score.intValue > restaurant2.score.intValue) {
+            return NSOrderedAscending;
+        }
+        return NSOrderedSame;
+    }];
+    return restaurantsToFilter;
+}
+
+- (NSNumber*)calculateRestaurantScore: (Restaurant *)restaurant{
+    int scalingFactor = 6;
+    int score = 0;
+    for (NSString *filter in self.filterPriority){
+        if([filter isEqualToString:@"price"]){
+            int priceScore = 0;
+            for(NSNumber *priceFilter in self.priceFilters){
+                if([priceFilter intValue] == [restaurant.priceValue intValue]){
+                    priceScore= 10*scalingFactor;
+                    break;
+                } else if ([priceFilter intValue] >  [restaurant.priceValue intValue]){
+                    priceScore = 5*scalingFactor;
+                }
+            }
+            score += priceScore;
+            
+        } else if ([filter isEqualToString:@"rating"]){
+            int ratingScore =0;
+            for (NSNumber *ratingFilter in self.ratingFilters){
+                if([ratingFilter intValue] == [restaurant.ratingValue intValue]){
+                    ratingScore = 10*scalingFactor;
+                    break;
+                } else if ([ratingFilter intValue] < [restaurant.ratingValue intValue]){
+                    ratingScore = 8*scalingFactor;
+                }
+            }
+            score += ratingScore;
+            
+        } else if ([filter isEqualToString:@"cuisine"]){
+            int cuisineScore=0;
+            NSUInteger perfectScore = self.cuisineFilters.count;
+            for (NSString *cuisineFilter in self.cuisineFilters){
+                for (NSDictionary *category in restaurant.categoriesArray){
+                    if([category[@"title"] isEqualToString:cuisineFilter]){
+                        cuisineScore++;
+                    }
+                }
+            }
+
+
+            if(cuisineScore == perfectScore){
+                score += (10*scalingFactor);
+            } else if (cuisineScore!=0 && cuisineScore > 10){
+                score += (8*scalingFactor);
+            } else {
+                score += (cuisineScore*scalingFactor);
+            }
+            
+        }
```

Comment:
This looks like another opportunity for some refactoring! Notice how all the if blocks are doing the same thing:
- looping through a filter array and getting a score by comparing against a restaurant's properties

This is slightly involved but here you could introduce a protocol for the filters like `FilterWithScoreType`, which would have a function to get the score of the filter so like this:

```
@protocol FilterWithScoreType <NSObject>

- (NSInteger)getScoreForRestaurant:(Restaurant *)restaurant;

@end
```

Then, all your filters can conform to this protocol and you can specify the scoring inside the protocol method implementation. Which would make your `calculateRestaurantScore` look like:

```
- (NSNumber *)calculateRestaurantScore:(Restaurant *)restaurant {
    int scalingFactor = 6;
    int score = 0;
    for (id<FilterWithScoreType filter in self.filterPriority) {
        score += [filter getScoreForRestaurant:restaurant];
    }
}
```

You would have to create classes for each filter, so this would be a pretty significant effort!